Code/Tools/AssetProcessor/native/utilities/BuilderList.cpp

```diff
@@ -0,0 +1,160 @@
+ *
+ */
+
+#include <native/utilities/BuilderList.h>
+
+namespace AssetProcessor
+{
+    void BuilderList::AddBuilder(AZStd::shared_ptr<Builder> builder, BuilderPurpose purpose)
+    {
+        if (purpose == BuilderPurpose::CreateJobs)
+        {
+            if (m_createJobsBuilder && m_createJobsBuilder->IsValid())
+            {
+                AZ_Error(
+                    "BuilderList", false, "AddBuilder called with CreateJobs builder (%s) but a CreateJobs builder (%s) already exists and is valid",
+                    builder->UuidString().c_str(), m_createJobsBuilder->UuidString().c_str());
+                return;
+            }
+
+            m_createJobsBuilder = AZStd::move(builder);
+        }
+        else
+        {
+            m_builders.emplace(builder->GetUuid(), builder);
+        }
+    }
+
+    AZStd::shared_ptr<Builder> BuilderList::Find(AZ::Uuid uuid)
+    {
+        if (m_createJobsBuilder && m_createJobsBuilder->GetUuid() == uuid)
+        {
+            return m_createJobsBuilder;
+        }
+
+        auto itr = m_builders.find(uuid);
+
+        return itr != m_builders.end() ? itr->second : nullptr;
+    }
+
+    BuilderRef BuilderList::GetFirst(BuilderPurpose purpose)
+    {
+        if (purpose == BuilderPurpose::CreateJobs)
+        {
+            if (m_createJobsBuilder)
+            {
+                if (!m_createJobsBuilder->m_busy)
+                {
+                    m_createJobsBuilder->PumpCommunicator();
+
+                    if (m_createJobsBuilder->IsValid())
+                    {
+                        return BuilderRef(m_createJobsBuilder);
+                    }
+
+                    m_createJobsBuilder = nullptr;
+                }
+                else
+                {
+                    AZ_Warning(
+                        "BuilderList",
+                        false,
+                        "CreateJobs builder requested but existing builder is already busy.  There should not be multiple parallel "
+                        "requests for CreateJobs builders");
+                }
+            }
+
+            return {};
+        }
+
+        for (auto itr = m_builders.begin(); itr != m_builders.end();)
+        {
+            auto& builder = itr->second;
+
+            if (!builder->m_busy)
+            {
+                builder->PumpCommunicator();
+
+                if (builder->IsValid())
+                {
+                    return BuilderRef(builder);
+                }
+
+                itr = m_builders.erase(itr);
+            }
+            else
+            {
+                ++itr;
+            }
+        }
+
+        return {};
+    }
+
+    AZStd::string BuilderList::RemoveByConnectionId(AZ::u32 connId)
+    {
+        AZStd::string uuidString;
+
+        if (m_createJobsBuilder && m_createJobsBuilder->GetConnectionId() == connId)
+        {
+            uuidString = m_createJobsBuilder->UuidString();
+            m_createJobsBuilder->m_connectionId = 0;
```

Comment:
Nit: I'm wondering whether it's necessary to set m_connectionId to 0 here and assume that the destructor for builder will reset all these member variables. Same for the builder->m_connectionId = 0 below.

Reply:
So this needs a comment.  Setting the connection Id to 0 basically marks the builder as invalid and disconnected.  The erase below isn't certain to destroy the builder.  If the builder is in-use, WaitForBuilderResponse will break out of its wait loop and fail the request as a result of this being set to 0.  Otherwise it would just continue on waiting for a response that will never happen.